build.assets/macos/tshdev/README.md

```diff
@@ -79,3 +81,104 @@ through the steps here unless you are trying to recreate parts of the skeleton
+going to make launchd start the program listed under `BundleProgram`. Extra arguments are passed to
+the program are under `ProgramArguments`. See `man launchd.plist` for more details behind individual
+plist properties.
+
+### Recompiling daemon
+
+Once a login item is enabled, you are free to update the binary advertised under `BundleProgram` as
+much as you want. Just remember to sign it before calling the daemon.
+
+### Modifying plist
+
+The situation is more complicated for modifying the plist itself. In theory, macOS should pick up on
+any updates made to the plist. In practice, the mechanism for that seems to be spotty, especially on
+development machines where apps are being rebuilt many times:
+
+* [Unable to register loginItem via SMAppService - Status Error 78](https://developer.apple.com/forums/thread/726826)
+* [Login Item failing to launch with SMAppService. Error: 78, LastExitStatus: 19968](https://forums.developer.apple.com/forums/thread/727785)
+
+When working on a prototype of the launch daemon, we've ran into a situation where macOS
+aggressively cached information from plists. Despite `BundleProgram` being changed to another path,
+macOS did not picked up on that.
+
+To see how macOS interprets the plist, there are two helpful commands:
+
+To dump information from the database of Launch Services (search for the identifier of the bundle or
+the label of the XPC service):
+
+```
+sfltool dumpbtm
+```
+
+To see how launchd interprets the plist:
+
+```
+launchctl print system/com.goteleport.tshdev.vnetd
+```
+
+If you find that macOS doesn't have up to date info about the plist, then…
+
+### Refreshing cached plist information
+
+The method we had the most success with involves these steps:
+
+1. If you made any changes inside `tsh.app`, like modifying plists, make sure to stage those changes
+   in git first.
+1. Close system settings if they're open.
+1. Move `tsh.app` to the trash and then either empty the trash or delete that specific item from the
+   trash.
+1. Open Login Items in system settings.
+
+This forces the Launch Services database to get updated. After opening Login Items, you might be
+able to see the old login item for tsh.app for a split second after which it's removed from the
+list.
+
+This completely resets the state of the login item, meaning that the daemon will have to be
+registered and enabled again. Now restore deleted files from tsh.app in git and then rebuild and
+sign the app. Hopefully, this time macOS will pick up any new changes.
+
+If that doesn't work, as a last resort there's a **very destructive** `sfltool resetbtm`. This
+basically wipes the whole database of Launch Services, meaning that the status of all login items
+gets reset – this includes other software installed on your machine.
+
+After running it, you must restart the device. We found it to be the most helpful when switching
+between local signed builds and tag builds. For some reason, in that scenario macOS wouldn't want to
```

Comment:
Could this be because of the change of .app location, or the fact that it gets confused because now multiple .apps for the same bundle exists?

I've run into weirdness with that before, too.

Reply:
In that case AFAIR I was putting tsh.app in the same location and there was no other bundles – I was doing that in a macOS VM. But maybe I did miss something. I couldn't find a good way to establish a link between a bundle identifier and the actual app bundle on disk that launchd is going to use for that particular login item, unfortunately.